packages/amazon-sumerian-hosts-core/src/core/awspack/LexFeature.js

```diff
@@ -0,0 +1,230 @@
+      );
+    }
+
+    // Add sumerian hosts user-agent
+    if (lexRuntime.config) {
+      lexRuntime.config.customUserAgent = Utils.withCustomUserAgent(
+        lexRuntime.config.customUserAgent
+      );
+    }
+
+    // Store parameters
+    this.SERVICES.lexRuntime = lexRuntime;
+  }
+
+  /**
+   * Sends audio user input to Amazon Lex.
+   *
+   * @param {Float32Array} inputAudio - TypedArray view of audio buffer
+   * @param {float} sourceSampleRate - Sample rate of input audio 
+   * @param {Object} config - Optional config for overriding lex bot info
+   * 
+   * @returns {Deferred}
+   */
+  processWithAudio(inputAudio, sourceSampleRate, config = {}) {
+    const audio = this._processAudio(inputAudio, sourceSampleRate);
+    return this._process('audio/x-l16; rate=16000', audio, config);
+  }
+
+  /**
+   * Sends text user input to Amazon Lex.
+   *
+   * @param {String} inputText - Text to send to lex bot
+   * @param {Object} config - Optional config for overriding lex bot info
+   * 
+   * @returns {Deferred}
+   */
+  processWithText(inputText, config = {}) {
+    return this._process('text/plain; charset=utf-8', inputText, config);
+  }
+
+  _process(contentType, inputStream, config) {
+    const settings = this._validateConfig(config);
+    const lexSettings = {
+      accept: 'text/plain; charset=utf-8',
+      botName: settings.botName,
+      botAlias: settings.botAlias,
+      contentType,
+      inputStream,
+      userId: settings.userId
+    };
+    return new Deferred((resolve, reject) => {
+      this.constructor.SERVICES.lexRuntime.postContent(lexSettings, (error, data) => {
+        if (error) {
+          return reject(error);
+        }
+        return resolve(data);
+      });
+    }).then(data => {
+      if (!data.message) {
+        if (data.encodedMessage) {
+          data.message = atob(data.encodedMessage);
+        }
+        else {
+          data.message = '';
+        }
+      }
+
+      this.emit(this.constructor.EVENTS.response, data.message);
+
+      return data.message;
+    });
+  }
+
+  _validateConfig(config) {
+    let settings = {};
+
+    settings.botName = config.botName ? config.botName : this._botName;
+    settings.botAlias = config.botAlias ? config.botAlias : this._botAlias;
+    settings.userId = config.userId ? config.userId : this._userId;
+
+    if (
+      settings.botName == undefined ||
+      settings.botAlias == undefined ||
+      settings.userId == undefined
+    )
+    {
+      throw new Error(
+        'Cannot process lex request. All arguments must be defined.'
+      );
+    }
+
+    return settings;
+  }
+
+  _processAudio(audioBuffer, sourceSampleRate) {
+    const downsampledAudio = LexUtils.downsampleAudio(audioBuffer, sourceSampleRate, this.constructor.LEX_DEFAULTS.SampleRate);
+    const encodedAudio = LexUtils.encodeWAV(downsampledAudio, this.constructor.LEX_DEFAULTS.SampleRate);
+    LexUtils.floatTo16BitPCM(encodedAudio, 44, downsampledAudio);
```

Comment:
Avoid magic numbers. 44 is a magic number. Explain to the reader why the value 44 is used.

If you don't know why 44 is used let's do some research. I know that there are two common sample rates for high-quality digital audio recording - 44.1 kHz and 48 kHz (kHz == thousands of samples per second of audio). Does the value 44 have something to do with that?

Reply:
44 is the offset after all the previous wav related config data in the buffer. Moved this function into the encode function

Reply:
While the 44 is now more hidden, the code does not tell the reader anymore about what '44' means than it did before. Kris's guess about its use being way off is a good signal that future team members/contributors are going to have no idea what this code is doing. Tools like constants and just comments of _why_ code is doing what it is doing will go a long way to keeping the code base maintainable.

Reply:
Right. I mean moving it to the encode function makes all those magic number easier to follow as well. Since from the function body, we can now know it just sets bunch of data of the buffer at starting offset 0 and all the way to 44.